storage/remote/intern.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Inspired / copied / modified from https://gitlab.com/cznic/strutil/blob/master/strutil.go,
+// which is MIT licensed, so:
+//
+// Copyright (c) 2014 The strutil Authors. All rights reserved.
+
+package remote
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+var interner = newPool()
+
+type pool struct {
+	mtx  sync.RWMutex
+	pool map[string]*entry
+}
+
+type entry struct {
+	s    string
+	refs int64
+}
+
+func newPool() *pool {
+	return &pool{
+		pool: map[string]*entry{},
+	}
+}
+
+func (p *pool) intern(s string) string {
+	if s == "" {
+		return ""
+	}
+
+	p.mtx.RLock()
+	interned, ok := p.pool[s]
+	p.mtx.RUnlock()
+	if ok {
+		atomic.AddInt64(&interned.refs, 1)
+		return interned.s
+	}
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+	if interned, ok := p.pool[s]; ok {
+		atomic.AddInt64(&interned.refs, 1)
+		return interned.s
+	}
+
+	s = pack(s)
+	p.pool[s] = &entry{
+		s:    s,
+		refs: 1,
+	}
+	return s
+}
+
+func (p *pool) release(s string) {
+	p.mtx.RLock()
+	interned, ok := p.pool[s]
+	p.mtx.RUnlock()
+
+	if !ok {
+		panic("released unknown string")
+	}
+
+	refs := atomic.AddInt64(&interned.refs, -1)
+	if refs > 0 {
+		return
+	}
+
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+	if atomic.LoadInt64(&interned.refs) != 0 {
+		return
+	}
+	delete(p.pool, s)
+}
+
+// StrPack returns a new instance of s which is tightly packed in memory.
+// It is intended for avoiding the situation where having a live reference
```

Comment:
Yeah these are never subsetted, so we could remove this.  I'm in favour of belt and braces just in case we start using this code elsewhere (such as in the scrape parser), but don't feel strongly.  Want me to remove it?

Reply:
The scrape parser has it's own thing already. I'd remove it for simplicity.